src/client/common/alias_dict.cpp

```diff
@@ -0,0 +1,213 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/cli/alias_dict.h>
+#include <multipass/constants.h>
+#include <multipass/standard_paths.h>
+
+#include <QDir>
+#include <QFile>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonParseError>
+#include <QTemporaryFile>
+
+#include "../../daemon/json_writer.h"
+
+mp::AliasDict::AliasDict()
+{
+    const auto file_name = QStringLiteral("%1_aliases.json").arg(mp::client_name);
+    const auto user_config_path = QDir{MP_STDPATHS.writableLocation(mp::StandardPaths::GenericConfigLocation)};
+    const auto cli_client_dir_path = QDir{user_config_path.absoluteFilePath(mp::client_name)};
+
+    aliases_file = cli_client_dir_path.absoluteFilePath(file_name).toStdString();
+
+    load_dict();
+}
+
+mp::AliasDict::~AliasDict()
+{
+    if (modified)
+    {
+        save_dict();
+    }
+}
+
+void mp::AliasDict::add_alias(const std::string& alias, const mp::AliasDefinition& command)
+{
+    if (aliases.try_emplace(alias, command).second)
+    {
+        modified = true;
+    }
+}
+
+bool mp::AliasDict::remove_alias(const std::string& alias)
+{
+    if (aliases.erase(alias) > 0)
+    {
+        modified = true;
+        return true;
+    }
+
+    return false;
+}
+
+mp::AliasDict::size_type mp::AliasDict::remove_aliases_for_instance(const std::string& instance)
+{
+    mp::AliasDict::size_type erased_elements = 0;
+
+    for (auto it = aliases.begin(); it != aliases.end();)
+    {
+        if (it->second.instance == instance)
+        {
+            ++erased_elements;
+            it = aliases.erase(it);
+        }
+        else
+            ++it;
+    }
+
+    if (erased_elements > 0)
+        modified = true;
+
+    return erased_elements;
+}
+
+mp::optional<mp::AliasDefinition> mp::AliasDict::get_alias(const std::string& alias) const
+{
+    try
+    {
+        return aliases.at(alias);
+    }
+    catch (const std::out_of_range&)
+    {
+        return mp::nullopt;
+    }
+}
+
+mp::AliasDict::DictType::iterator mp::AliasDict::begin()
+{
+    return aliases.begin();
+}
+
+mp::AliasDict::DictType::iterator mp::AliasDict::end()
+{
+    return aliases.end();
+}
+
+mp::AliasDict::DictType::const_iterator mp::AliasDict::cbegin() const
+{
+    return aliases.cbegin();
+}
+
+mp::AliasDict::DictType::const_iterator mp::AliasDict::cend() const
+{
+    return aliases.cend();
+}
+
+bool mp::AliasDict::empty() const
+{
+    return aliases.empty();
+}
+
+mp::AliasDict::size_type mp::AliasDict::size() const
+{
+    return aliases.size();
+}
+
+void mp::AliasDict::load_dict()
+{
+    QFile db_file{QString::fromStdString(aliases_file)};
+
+    aliases.clear();
+
+    if (!db_file.open(QIODevice::ReadOnly))
```

Comment:
I wonder if there should be more of an error here is the dictionary database cannot be opened?

Reply:
If the database file does not exist, we can safely `return` because there is nothing to do within the function. I added an error in case the file exists and cannot be opened, which I didn't consider before your comment.